datalad/interface/tests/test_ls.py

```diff
@@ -58,6 +64,162 @@ def test_ls_repos(toppath):
+        # running in reverse order (non-recursive followed by recursive mode) will give (0, actual size)
+        assert_equal(child['size']['total'], '6 Bytes')
+
+        # verify subdirectory traversal if run in recursive mode
+        if recursive:
+            # sub-dictionary should not include git and hidden directory info
+            assert_equal([item for item in child['nodes'] if ('subgit' or '.fgit') == item['name']], [])
+            # extract subdirectory dictionary, else fail
+            subchild = [subitem for subitem in child["nodes"] if subitem['name'] == 'subdir'][0]
+            # extract info of file1.txts, else fail
+            link = [subnode for subnode in subchild["nodes"] if subnode['name'] == 'file1.txt'][0]
+            # verify node's sizes and type
+            assert_equal(link['size']['total'], '3 Bytes')
+            assert_equal(link['size']['ondisk'], link['size']['total'])
+            assert_equal(link['type'], 'link')
+            # extract info of file2.txt, else fail
+            brokenlink = [subnode for subnode in subchild["nodes"] if subnode['name'] == 'file2.txt'][0]
+            # verify node's sizes and type
+            assert_equal(brokenlink['type'], 'link-broken')
+            assert_equal(brokenlink['size']['ondisk'], '0 Bytes')
+            assert_equal(brokenlink['size']['total'], '3 Bytes')
+
+
+@with_tree(
+    tree={'dir': {'.fgit': {'ab.txt': '123'},
+                  'subdir': {'file1.txt': '123', 'file2.txt': '123'},
+                  'subgit': {'fgit.txt': '123'}},
+          '.hidden': {'.hidden_file': '123'}})
+def test_ls_json(topdir):
+    annex = AnnexRepo(topdir, create=True)
+    ds = Dataset(topdir)
+    # create some file and commit it
+    open(opj(ds.path, 'subdsfile.txt'), 'w').write('123')
+    ds.install(path='subdsfile.txt')
+    ds.save("Hello!", version_tag=1)
+    # add a subdataset
+    ds.install('subds', source=topdir)
+
+    git = GitRepo(opj(topdir, 'dir', 'subgit'), create=True)                    # create git repo
+    git.add(opj(topdir, 'dir', 'subgit', 'fgit.txt'), commit=True)              # commit to git to init git repo
+    annex.add(opj(topdir, 'dir', 'subgit'), commit=True)                        # add the non-dataset git repo to annex
+    annex.add(opj(topdir, 'dir'), commit=True)                                  # add to annex (links)
+    annex.drop(opj(topdir, 'dir', 'subdir', 'file2.txt'), options=['--force'])  # broken-link
+
+    meta_dir = '.git/datalad/metadata'
+    meta_path = opj(topdir, meta_dir)
+
+    for all in [True, False]:
+        for recursive in [True, False]:
+            for state in ['file', 'delete']:
+                with swallow_logs(), swallow_outputs():
+                    ds = _ls_json(topdir, json=state, all=all, recursive=recursive)
+
+                # subdataset should have its json created and deleted when all=True else not
+                subds_metahash = hashlib.md5('/'.encode('utf-8')).hexdigest()
```

Comment:
duplication is evil! ;)  make a function `get_metahash` or smth like that, which would get your path, return a hash.  Use here and elsewhere